### PR TITLE
Add T5 e2e_benchmark translation task with support for DDP 

### PR DESCRIPTION
**Pull request by erichan1**
What this benchmark will have
- Add T5 e2e benchmark
- Allow DDP and Deepspeed to run on T5 
- Have both train and eval only modes, and include eval as part of train
- Add new wmt16 dataset for general translation tasks. Can be extended to other seq2seq models. 

To Do Later
- Add FSDP

### CODE REVIEW

File: torchbenchmark/e2e_models/hf_t5/__init__.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+        # We need to recalculate our total training steps as the size of the training dataloader may have changed.
+        num_update_steps_per_epoch = math.ceil(len(train_dataloader) / hf_args.gradient_accumulation_steps)
+        if overrode_max_train_steps:
+            hf_args.max_train_steps = hf_args.num_train_epochs * num_update_steps_per_epoch
+        # Afterwards we recalculate our number of training epochs
+        hf_args.num_train_epochs = math.ceil(hf_args.max_train_steps / num_update_steps_per_epoch)
+        # TODO: not doing checkpointing
+        # # Figure out how many steps we should save the Accelerator states
+        # if hasattr(hf_args.checkpointing_steps, "isdigit"):
```
Comment (erichan1) [inline torchbenchmark/e2e_models/hf_t5/__init__.py:255]: Commented out checkpointing. Do we want to include this? We'll have to save the checkpoints somewhere.

File: torchbenchmark/util/framework/transformers/translation/args.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
+"""
+Hacked from https://github.com/huggingface/transformers/blob/main/examples/pytorch/translation/run_translation_no_trainer.py 
+It runs HuggingFace transformer models translation on WMT16
+"""
+import argparse
+from transformers import SchedulerType
+
+task_to_keys = {
```
Comment (erichan1) [inline torchbenchmark/util/framework/transformers/translation/args.py:8]: I separate the dataset specific args like this. See init py to see how tasks_to_keys is used. Not sure if this is the best way to separate dataset/task specific hf_args and model-specific hf_args.

File: torchbenchmark/e2e_models/hf_t5/__init__.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        task_name = self.tb_args.task_name
+        task_args = task_to_keys[task_name] # dataset specific hf_args
+        # T5 requires source prefix to know what to translate
+        if task_name == "wmt-en-ro":
+            source_prefix = "translate English to Romanian: "
+        elif task_name == "wmt-en-de":
+            source_prefix = "translate English to German: "
+        else:
+            raise RuntimeError(f"Unsupported translation task {task_name} for model hf_t5")
+        task_args.extend(["--source_prefix", source_prefix])
```
Comment (erichan1) [inline torchbenchmark/e2e_models/hf_t5/__init__.py:64]: Counterpart to [comment](https://github.com/pytorch/benchmark/pull/1122/files#r955279905) below - is this the best abstraction to separate translation task specific args and model specific args?

**Review by wconstab — APPROVED** [review]

**Comment by erichan1** [comment]
Example output:

```
{
    "name": "distributed",
    "environ": {
        "pytorch_git_version": "1edaa546e8d736998925e5f29545443729e8d40e"
    },
    "args": {
        "ngpus": 8,
        "nodes": 1,
        "timeout": 1440,
        "profiler": false,
        "partition": "train",
        "cluster": null,
        "job_dir": "/data/home/erichan1/cluster/work/benchmark/.userbenchmark/distributed/e2e_hf_t5",
        "model": "torchbenchmark.e2e_models.hf_t5.Model",
        "trainer": "torchbenchmark.util.distributed.trainer.Trainer",
        "distributed": "none",
        "dist_url": "file:///data/home/erichan1/cluster/work/benchmark/.userbenchmark/distributed/e2e_hf_t5/81889640a97741f39bc0e09a35a8ebbc_init",
        "output_dir": "/data/home/erichan1/cluster/work/benchmark/.userbenchmark/distributed/e2e_hf_t5",
        "extra_args": []
    },
    "metrics": {
        "0-fwd_mean": 41.60278739929199,
        "0-fwd_stdev": 0.4013525019877281,
        "0-bwd_mean": 95.8198974609375,
        "0-bwd_stdev": 2.074533412481593,
        "0-opt_mean": 21.429196739196776,
        "0-opt_stdev": 0.17533490658203918,
        "1-fwd_mean": 42.76195793151855,
        "1-fwd_stdev": 1.4134329483327561,
        "1-bwd_mean": 92.92210235595704,
        "1-bwd_stdev": 1.8270807455096747,
        "1-opt_mean": 23.3517183303833,
        "1-opt_stdev": 0.7375556447869801,
        "2-fwd_mean": 44.67186584472656,
        "2-fwd_stdev": 1.7146957380977343,
        "2-bwd_mean": 91.51677474975585,
        "2-bwd_stdev": 2.896919213453732,
        "2-opt_mean": 22.664323234558104,
        "2-opt_stdev": 1.86678946881834,
        "3-fwd_mean": 40.50614433288574,
        "3-fwd_stdev": 1.0314747922210197,
        "3-bwd_mean": 96.58173828125,
        "3-bwd_stdev": 2.184991773765833,
        "3-opt_mean": 21.743779182434082,
        "3-opt_stdev": 0.15668476415387397,
        "4-fwd_mean": 59.395945358276364,
        "4-fwd_stdev": 1.6227717010554719,
        "4-bwd_mean": 68.6760971069336,
        "4-bwd_stdev": 1.2091913189048988,
        "4-opt_mean": 30.541087913513184,
        "4-opt_stdev": 0.962546673438117,
        "5-fwd_mean": 40.876163482666016,
        "5-fwd_stdev": 0.1440217928959673,
        "5-bwd_mean": 94.09861831665039,
        "5-bwd_stdev": 2.1851929379312893,
        "5-opt_mean": 23.94410228729248,
        "5-opt_stdev": 0.19617251586344805,
        "6-fwd_mean": 40.67866249084473,
        "6-fwd_stdev": 0.13880394008073413,
        "6-bwd_mean": 96.64809036254883,
        "6-bwd_stdev": 2.1530119827625,
        "6-opt_mean": 21.548608207702635,
        "6-opt_stdev": 0.05699353177171309,
        "7-fwd_mean": 41.827459716796874,
        "7-fwd_stdev": 0.18507004474107516,
        "7-bwd_mean": 95.24132385253907,
        "7-bwd_stdev": 2.728420048685656,
        "7-opt_mean": 21.826060676574706,
        "7-opt_stdev": 0.14197387907529965
    }
}
```

File: torchbenchmark/util/framework/transformers/translation/args.py (hunk 2 above)
Comment (xuzhao9) [reply torchbenchmark/util/framework/transformers/translation/args.py:8]: Yep it looks good to me,

**Review by xuzhao9 — APPROVED** [review]
LGTM!

File: torchbenchmark/e2e_models/hf_t5/__init__.py (hunk 3 above)
Comment (xuzhao9) [reply torchbenchmark/e2e_models/hf_t5/__init__.py:64]: Yes, I think it is a good abstraction.

**Comment by xuzhao9** [comment]
@erichan1 Can we have the Deepspeed result here as well?

**Comment by erichan1** [comment]
> @erichan1 Can we have the Deepspeed result here as well?

I cannot get it to work on AWS cluster because of slurm. Land this for now and fix deepspeed in the next one? I also cannot run hf_bert deepspeed for the same reason.

**Comment by xuzhao9** [comment]
> > @erichan1 Can we have the Deepspeed result here as well?
> 
> I cannot get it to work on AWS cluster because of slurm. Land this for now and fix deepspeed in the next one? I also cannot run hf_bert deepspeed for the same reason.

Got it, let's land this and re-visit deepspeed issue later.

**Pull request closed without merging**